### PR TITLE
Adding REST API to support displayOnDevportal change for deployments

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**
## Issues

adding rest api related fix for : https://github.com/wso2/product-apim/issues/10281

## Implementation

- Two new REST APIs were added for API and API Products

**PUT** https://localhost:9443/api/am/publisher/v2/apis/{API_ID}/deployments/{deploymentId}

**PUT** https://localhost:9443/api/am/publisher/v2/api-products/{API_PRODUCT_ID}/deployments/{deploymentId}

**Sample Request**

PUT https://localhost:9443/api/am/publisher/v2/apis/35611d64-d820-4a1d-b4ab-e21ee9833466/deployments/UHJvZHVjdGlvbiBhbmQgU2FuZGJveA

PUT https://localhost:9443/api/am/publisher/v2/api-products/b90917a6-ca86-4223-a38a-d6cc86f11a63/deployments/UHJvZHVjdGlvbiBhbmQgU2FuZGJveA

**Sample Payload**

{
    "revisionUuid": "063a587e-2fbc-449e-b5e4-e4971beb6b9c",
    "displayOnDevportal": false
}

- {deploymentId} is the **base64 url encoded** value of the environment name

- Altered GET /deployed-revision  to GET /deployments

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 1]
```diff
@@ -16260,6 +16260,41 @@ public void removeAPIRevisionDeployment(String apiUUID, Set<APIRevisionDeploymen
...
+                PreparedStatement statement = connection
+                        .prepareStatement(SQLConstants.APIRevisionSqlConstants.UPDATE_API_REVISION_DEPLOYMENT_MAPPING);
+                for (APIRevisionDeployment deployment : deployments) {
+                    statement.setBoolean(1, deployment.isDisplayOnDevportal());
+                    statement.setString(2, deployment.getDeployment());
+                    statement.setString(3, deployment.getRevisionUUID());
+                    statement.addBatch();
+                }
+                statement.executeBatch();
+                connection.commit();
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml [hunk 2]
```diff
@@ -6584,7 +6626,49 @@ paths:
...
+        200:
+          description: |
+            Created.
+            Successful response with the newly updated APIRevisionDeployment List object as the entity in the body.
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/APIRevisionDeployment'
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml:6661]: we need to return only the updated single deployment.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java [hunk 3]
```diff
@@ -4881,4 +4882,25 @@ private APIDTO importAsyncAPISpecification(InputStream definition, String defini
         }
         return null;
     }
+
+    @Override
+    public Response changeAPIDisplayOnDevportal(String apiId, String deploymentId, APIRevisionDeploymentDTO apIRevisionDeploymentDTO,
+                                                MessageContext messageContext) throws APIManagementException {
+        APIProvider apiProvider = RestApiCommonUtil.getLoggedInUserProvider();
+        String revisionId = apIRevisionDeploymentDTO.getRevisionUuid();
+        String decodedDeploymentName = new String(Base64.getUrlDecoder().decode(deploymentId), StandardCharsets.UTF_8);
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java [hunk 4]
```diff
@@ -4881,4 +4882,25 @@ private APIDTO importAsyncAPISpecification(InputStream definition, String defini
...
+        apiRevisionDeployment.setVhost(apIRevisionDeploymentDTO.getVhost());
+        apiRevisionDeployment.setDisplayOnDevportal(apIRevisionDeploymentDTO.isDisplayOnDevportal());
+        apiProvider.updateAPIDisplayOnDevportal(apiId, revisionId, apiRevisionDeployment);
+        List<APIRevisionDeployment> apiRevisionDeploymentsResponse = apiProvider.getAPIRevisionDeploymentList(revisionId);
+        List<APIRevisionDeploymentDTO> apiRevisionDeploymentDTOS = new ArrayList<>();
+        for (APIRevisionDeployment apiRevisionDeploymentGet : apiRevisionDeploymentsResponse) {
+            apiRevisionDeploymentDTOS.add(APIMappingUtil.fromAPIRevisionDeploymenttoDTO(apiRevisionDeploymentGet));
+        }
+        Response.Status status = Response.Status.CREATED;
+        return Response.status(status).entity(apiRevisionDeploymentDTOS).build();
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml (hunk 2 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml [hunk 5]
```diff
@@ -1588,7 +1588,49 @@ paths:
...
+    #--------------------------------------------
+    # Change Display On Devportal Field
+    #--------------------------------------------
+    put:
+      tags:
+        - API Revisions
+      summary: Update Deployment
+      description: |
+        Update display on devportal field in a deployment
+      operationId: changeAPIDisplayOnDevportal
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml [hunk 6]
```diff
@@ -1588,7 +1588,49 @@ paths:
...
+        200:
+          description: |
+            Created.
+            Successful response with the newly updated APIRevisionDeployment List object as the entity in the body.
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/APIRevisionDeployment'
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml]: resolved

**Review by Krishanx92 — APPROVED** [review]

**Review by malinthaprasan — APPROVED** [review]

**Pull request merged**